tests/util/bft.py

```diff
@@ -289,60 +289,87 @@ def get_live_replicas(self):
+            print(f'View #{matching_view} has been activated by '
+                  f'{nb_replicas_in_matching_view} >= n-f = {self.config.n - self.config.f}')
+
+            return matching_view
         except trio.TooSlowError:
             assert False, err_msg + \
-                          f'(lastAgreedView={last_agreed_view} ' \
-                          f'replicasInExpectedView={replicas_in_expected_view})'
+                          f'(matchingView={matching_view} ' \
+                          f'replicasInMatchingView={nb_replicas_in_matching_view})'
+
+    async def _wait_for_matching_agreed_view(self, replica_id, expected):
+        """
+        Wait for the last agreed view to match the "expected" predicate
+        """
+        last_agreed_view = None
+        with trio.fail_after(seconds=30):
+            while True:
+                try:
+                    with trio.move_on_after(seconds=1):
+                        key = ['replica', 'Gauges', 'lastAgreedView']
+                        view = await self.metrics.get(replica_id, *key)
+                        if expected(view):
+                            last_agreed_view = view
+                            break
+                except KeyError:
+                    # metrics not yet available, continue looping
+                    continue
+        return last_agreed_view
+
+    async def _wait_for_active_view(self, view):
+        """
+        Wait for a view to become active on enough (n-f) replicas
+        """
+        with trio.fail_after(seconds=30):
+            while True:
+                nb_replicas_in_view = await self._count_replicas_in_view(view)
+
+                # wait for n-f = 2f+2c+1 replicas to be in the expected view
+                if nb_replicas_in_view >= 2 * self.config.f + 2 * self.config.c + 1:
+                    break
+        return nb_replicas_in_view
+
+    async def _count_replicas_in_view(self, view):
+        """
+        Count the number of replicas that have activated a given view
+        """
+        nb_replicas_in_view = 0
+
+        async def count_if_replica_in_view(r, expected_view):
+            """
+            A closure that allows concurrent counting of replicas
+            that have activated a given view.
+            """
+            nonlocal nb_replicas_in_view
+
+            key = ['replica', 'Gauges', 'currentActiveView']
+
+            with trio.move_on_after(seconds=5):
+                while True:
+                    with trio.move_on_after(seconds=1):
+                        try:
+                            replica_view = await self.metrics.get(r, *key)
+                            if replica_view == expected_view:
+                                nb_replicas_in_view += 1
+                        except KeyError:
+                            # metrics not yet available, continue looping
+                            continue
+                        else:
+                            break
```

Comment:
I believe you meant to do this, since there's another loop in `_wait_for_active_view` that retries, but I want to make sure. This will break and not retry if `replica_view != expected_view`. However, other replicas may still try for 5 seconds if they are unavailable... We could take that time to retry in this loop. I don't think there's a tradeoff here, since even if all replicas are available and get the non matching view they would just retry in this loop vs the `wait_for_active_view` loop.

Either way, it's not a big deal and the code is correct. I think my suggestion may be an optimization in future commits though.

Reply:
Thanks Andrew, we can indeed address this as a separate optimization.